KAFKA-10000: Exactly-once support for source connectors (KIP-618)

C0urante opened the pull request:
Implements [KIP-618](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors#KIP618:ExactlyOnceSupportforSourceConnectors-Newmetrics).

Depends on changes from https://github.com/apache/kafka/pull/11046

There are several changes here that can be reviewed fairly independently of each other:
- Support for transactional source tasks, which is largely implemented in the `ExactlyOnceWorkerSourceTask` class, its newly-introduced `AbstractWorkerSourceTask` superclass, the `Worker` class (whose API for task starts has been split up from the existing `startTask` method into separate `startSinkTask`, `startSourceTask`, and `startExactlyOnceSourceTask` methods), and the `WorkerTransactionContext` class (which is used to allow connectors to define their own transaction boundaries)
- Zombie fencing logic and the use of a transactional producer for some writes to the config topic, which are done by the leader of the cluster and are largely implemented in the `DistributedHerder`, `ConfigBackingStore`, `KafkaConfigBackingStore`, and `ClusterConfigState` classes
- A new method in the `Admin` API for fencing out transactional producers by ID, which is done with changes to the `Admin` interface (unsurprisingly) and the `KafkaAdminClient` class
- Support for per-connector offsets topics, which touches on the `Worker`, `OffsetStorageReaderImpl`, and `OffsetStorageWriter` classes
- A few new `SourceConnector` methods for communicating support for exactly-once guarantees and connector-defined transactions; these take place in the `SourceConnector` class (also unsurprisingly) and the `AbstractHerder` class

Existing unit tests are expanded where applicable, and new ones have been introduced where necessary.

Eight new integration tests are added, which cover scenarios including preflight validation checks, all three types of transaction boundary, graceful recovery of the leader when fenced out from the config topic by another worker, ensuring that the correct number of task producers are fenced out across generations, accurate reporting of failure to bring up tasks when fencing does not succeed (includes an ACL-secured embedded Kafka cluster to simulate one of the most likely potential causes of this issue--insufficient permissions on the targeted Kafka cluster), and the use of a custom offsets topic.

Many but not all existing system tests are modified to add cases involving exactly-once source support, which helps give us reasonable confidence that the feature is agnostic with regards to rebalance protocol. A new test is added that is based on the existing bounce test, but with no sink connector and with stricter expectations for delivery guarantees (no duplicates are permitted).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante commented:
@gharris1727 you've been tremendously helpful so far in the design process; if you have time would you like to take a look at the implementation as well?

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -337,15 +353,29 @@ private void readToLogEnd() {
                     log.trace("Behind end offset {} for {}; last-read offset is {}",
                             endOffset, topicPartition, lastConsumedOffset);
-                    poll(Integer.MAX_VALUE);
+                    if (topicContainsTransactions) {
+                        // The consumer won't return from its poll method if a transaction is aborted, even though
+                        // its position will advance. So, we poll for at most one second, then give ourselves another
+                        // chance to check whether we've reached the end of the topic.
+                        poll(1000);
```
This causes a new consumer fetch request to be issued every second while there's an active read-to-end request for the log.

If this is too expensive and ends up causing issues (either on the worker or on the broker), we might be able to retain the existing `poll(Integer.MAX_VALUE)` behavior by first producing a sentinel message to each topic partition we want to read to the end of. That way, even if a transaction that was initially blocking our read to the end of the partition is aborted, the call to `Consumer::poll` will return, since there will be a non-transactional message after it.

Existing, older workers would be able to handle these sentinel messages gracefully (the only downside would be some `ERROR`-level log messages from the `KafkaConfigBackingStore` class upon encountering unrecognized in the config topic). However, we'd need to reason carefully about how this might work with multiple workers trying to read to the end of the same topic partitions at the same time.

C0urante replied:
I've filed https://issues.apache.org/jira/browse/KAFKA-12980 and plan to begin work on it. Hopefully we can leverage that improvement here to avoid the `poll(1000)` hack.

rhauch replied:
And for reference, here is @C0urante's PR for that fix: #11046

I agree, it would be good to avoid this hack if possible.

C0urante replied:
Now that https://github.com/apache/kafka/pull/11046 has been merged and this PR has been rebased to include it, we can resolve this comment.

C0urante commented:
There was a small misunderstanding of the rebalance logic for distributed workers. It turns out that they already preemptively stop reconfigured tasks before (re)joining the group, so no additional logic is necessary to stop reconfigured source tasks before they are naturally fenced out by the leader. The KIP has been updated with this note and the changes in the PR related to it have been reverted.

C0urante committed:
KAFKA-10000: Exactly-once support for source connectors

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceTaskContext.java:
```diff
@@ -38,4 +38,30 @@
+     * </pre>
+     *
+     * @return the transaction context, or null if the user does not want the connector to define
+     * its own transaction boundaries
```
Nit: what does "user" mean here? Should we instead refer to the connector configuration, perhaps something like: "or null if the connector was configured to not specify transaction boundaries"?

(I know this is taken directly from the KIP, but I think we should improve this JavaDoc during the implementation.)

C0urante replied:
No objections to modifying Javadocs here instead of in a KIP (especially since you made that clear on the mailing list). It's easier to review these sorts of details in a PR IMO anyways.

Can make the change to refer to the connector configuration. Agree that "user" is ambiguous.

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java:
```diff
@@ -28,4 +30,31 @@
+    }
+
+    /**
+     * Signals whether the connector can define its own transaction boundaries with the proposed
+     * configuration. Developers must override this method if they wish to add connector-defined
+     * transaction boundary support; if they do not, users will be unable to create instances of
+     * this connector that use connector-defined transaction boundaries. The default implementation
+     * will return {@code UNSUPPORTED}.
```
I think this JavaDoc should more clearly specify when an implementation may expect this method is called relative to other methods. We don't want to box ourselves in with respect to the implementation, but we also need a clear contract that Connector developers can rely upon.

For example, maybe add something like: 
> This method may be called by the runtime before the {@link #start} method when the connector is being run with exactly-once support and when the connector is to determine the transactions boundaries.

C0urante replied:
Can do. I think a reference to the existing `validate` method may help clarify things (especially since this method will be used in almost exactly the same way); LMK what you think.

rhauch replied:
WDYT about something like this:
```
/**
 * Signals whether the connector implementation is capable of defining the transaction boundaries for a
 * connector with the given configuration. This method is called before {@link #start(Map)}, only when the
 * runtime supports exactly-once and the connector configuration includes {@code transaction.boundary=connector}.
 *
 * <p>This method need not be implemented if the connector implementation does not support definiting
 * transaction boundaries.
 *
 * @param connectorConfig the configuration that will be used for the connector
 * @return {@link ConnectorTransactionBoundaries.SUPPORTED} if the connector will define its own transaction boundaries, 
 *         or {@link ConnectorTransactionBoundaries.UNSUPPORTED} otherwise.
 * @see TransactionContext
 */
```

C0urante replied:
LGTM. I added a small note on not returning `null` with the same language as in `SourceConnector::exactlyOnceSupport` but otherwise added this verbatim. LMKWYT

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java:
```diff
@@ -28,4 +30,31 @@
+    /**
+     * Signals whether the connector supports exactly-once delivery guarantees with a proposed configuration.
+     * Developers can assume that worker-level exactly-once support is enabled when this method is invoked.
+     * The default implementation will return {@code null}.
```
I think this JavaDoc should more clearly specify when an implementation may expect this method is called relative to other methods. We don't want to box ourselves in with respect to the implementation, but we also need a clear contract that Connector developers can rely upon.

For example, maybe add something like: 
> This method may be called by the runtime before the {@link #start} method when the connector is being run with exactly-once support.

C0urante replied:
Ack, done.

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java:
```diff
@@ -28,4 +30,31 @@
+     * connector cannot.
+     */
+    public ExactlyOnceSupport exactlyOnceSupport(Map<String, String> connectorConfig) {
+        return null;
```
It's difficult to discern in the KIP, but IIUC there is a difference between this method returning `null` versus `ExactlyOnceSupport#UNSUPPORTED`:
* Returning `null` really seems to equate to `UNKNOWN` and a Connect user is still allowed sort of force the use of EOS by setting the `exactly.once.support=requested` in the connector configuration.
* Returning `UNSUPPORTED` means that a Connect user is not allowed to force the use of EOS.

Should we just have another enum for `UNKNOWN` and make this more explicit than "null"?

Also, it seems like it would make sense to document that this method should be overridden by Connector developers, but has a default for backward compatibility. And it should state more clearly what should be returned for the various options.

C0urante replied:
> Should we just have another enum for UNKNOWN and make this more explicit than "null"?

This was actually [suggested in the discussion thread](https://mail-archives.apache.org/mod_mbox/kafka-dev/202105.mbox/%3cCAMdOrUX4CvPsb+yjfTenHyRTtE=2aaw-_-_b2VbD+pvqzY7bxw@mail.gmail.com%3e):

> what do you think about a new "exactlyOnce()" method to the SourceConnector class that can return a new ExactlyOnce enum with options of "SUPPORTED", "UNSUPPORTED", and "UNKNOWN", with a default implementation that returns "UNKNOWN"?

And [decided against](https://mail-archives.apache.org/mod_mbox/kafka-dev/202105.mbox/%3cCADXUNmbsypos0LeJ8KXW9EAPCxC7wbtGXQDQHrpu6qRbJWiaYQ@mail.gmail.com%3e):

> The problem with having an explicit UNKNOWN case is we really want connector developers to _not_ use it. That could mean it's deprecated from the start. Alternatively we could omit it from the enum and use null to mean unknown (we'd have to check for a null result anyway), with the contract for the method being that it should return non-null. Of course, this doesn't remove the ambiguous case, but avoids the need to eventually remove UNKNOWN in the future.

(What I found especially convincing in the snippet above were the points that 1) we don't want people to return `UNKNOWN` from this method, and 2) no matter what, we're going to have to check for `null` anyways.)


> Also, it seems like it would make sense to document that this method should be overridden by Connector developers, but has a default for backward compatibility.

Ack, can do.

> And it should state more clearly what should be returned for the various options.

I've taken a shot at this, not sure how much clearer it can get but if you have thoughts let me know.

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceTask.java:
```diff
@@ -20,13 +20,45 @@
+     */
+    public static final String TRANSACTION_BOUNDARY_CONFIG = "transaction.boundary";
+
+    public enum TransactionBoundary {
```
The `SourceTask` class is part of the public API for Connect, and so we should have JavaDoc on this enum and its literals and methods.

C0urante replied:
Ack, done.

C0urante replied:
Just making sure: is there anything left to address for this comment?

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceTask.java:
```diff
@@ -20,13 +20,45 @@
+
+        public static final TransactionBoundary DEFAULT = POLL;
+
+        public static List<String> options() {
```
Do we really need this method when `values()` is already available and more standard? It looks like this is used in only two places: one expects an Array (where `values()` would work better), and the other just uses it to build a documentation (where `Arrays.toString(...)` might work just as well).

C0urante replied:
I wanted a convenient way to bring everything to lowercase, which is more standard for properties like this (see how [values for the consumer `isolation.level` property are rendered](https://github.com/apache/kafka/blob/da38a1df273ec9d3a077435b2a63d75053edd308/clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java#L557), for example) and, IMO, more readable.

We could remove this method and replace it with inline calls to `values()` followed by some streams magic to lowercase at the call site, but that seemed less clean than this approach.

Alternatively, we could introduce a new utility method (such as `Utils::enumNames`) that does this for us in a centralized, reusable location and obviates the need for public-API methods that may cause headaches down the road.

Thoughts?

rhauch replied:
I think the risk of introducing `options()` is that some developers might accidentally use `values()`. 

The pattern used in `ConnectorType` is far better, as it overrides the `toString()` method. That doesn't handle the case-independence for parsing, though `ConverterType` is a better pattern to follow if that's required.

Let's be consistent with the new enums, and have each follow one of those two patterns depending upon whether parsing case-independently is required.

C0urante replied:
Neither of these patterns play nicely with the validators offered by `ConfigDef`, though. Both `ValidString` and `CaseInsensitiveValidString` require an array (or really, a varargs list) of strings in their constructors, which prevents us from using `Enum::values` directly to create one of these validators.

I've removed `options()` (from both `TransactionBoundary` and `ExactlyOnceSupportLevel`) in favor of a reusable `Utils::enumOptions` method, which returns a `String[]` containing the names (retrieved via `toString`) of each value for an enum class. This prevents us from having to duplicate this logic and allows us to use it without adding new methods to public API.

The actual implementations for both enums now follow the `ConnectorType` pattern, with an overridden `toString` method that returns the lowercase name of the enum and a static `fromProperty` method (which mirrors the `ConnectorType::fromValue` method) that parses a string case-insensitively into a value for the enum. I opted for a different method name for the latter since the values given to `fromProperty` should always be user-supplied values and the motivation for this separate method (instead of just invoking `Enum::valueOf` directly) is directly related to parsing user-supplied values.

We could also add a general-purpose `EnumValidator` class to `ConfigDef` (for public use) or in a section of the Connect code base reserved for private API, but that would still leave us with the trouble of having to re-implement logic to follow the phrase "Permitted values are" in property docstrings, and it's probably best to hold off on implementing a public-facing validator like that until its exact behavior can be agreed on with a KIP.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedConfig.java:
```diff
@@ -192,6 +198,19 @@
+            + "and that some source connectors may still not be able to provide exactly-once delivery guarantees even with this support enabled. "
+            + "Permitted values are \"disabled\", \"preparing\", and \"enabled\". In order to safely enable exactly-once support for source connectors, "
+            + "all workers in the cluster must first be updated to use the \"preparing\" value for this property. "
+            + "Once this has been done, a second update of all of the workers in the cluster should be performed to change the value of this property to \"enabled\".";
```
We should mention in the public docs that when enabling exactly once support for source connectors, consumers of the topics to which the EOS source connectors write should be configured with `isolation.level=read_committed`? After all, the default for consumers is `isolation.level=read_uncommitted` (see [ConsumerConfig.java](https://github.com/apache/kafka/blob/db1f581da7f3440cfd5be93800b4a9a2d7327a35/clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java#L302)).

It's clear to me that we should mention this, but it's not clear where we should do so. The user documentation generated from this doc string might be one spot that users will see routinely, so maybe it's a candidate. Another would be in the Kafka Connect docs about EOS for source connectors.

BTW, can you add to this PR changes to the Kafka docs that describe this feature?

C0urante replied:
> It's clear to me that we should mention this, but it's not clear where we should do so. The user documentation generated from this doc string might be one spot that users will see routinely, so maybe it's a candidate. Another would be in the Kafka Connect docs about EOS for source connectors.

I think both are acceptable. I'll add this to the docstring now and include it in the high-level docs when I write those as well.

> BTW, can you add to this PR changes to the Kafka docs that describe this feature?

Given how massive this PR is already, I'd like to do this in a follow-up, with the understanding that Kafka docs changes are a requisite for including this feature in a release.

C0urante replied:
I've unresolved this as a general reminder to make sure to write the docs section for this feature before we release it.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedConfig.java:
```diff
@@ -401,10 +429,46 @@ public Integer getRebalanceTimeout() {
+    public boolean transactionalLeaderEnabled() {
+        return Arrays.asList(EXACTLY_ONCE_SOURCE_SUPPORT_ENABLED, EXACTLY_ONCE_SOURCE_SUPPORT_PREPARING)
+                .contains(getString(EXACTLY_ONCE_SOURCE_SUPPORT_CONFIG).toLowerCase(Locale.ROOT));
+    }
```
Should we have an enum for the `enabled`, `preparing` and `disabled` literals, and should these make use of them? Also, it might be useful to have JavaDoc on these methods, simply to help future developers understand the intent.

C0urante replied:
> Should we have an enum for the enabled, preparing and disabled literals, and should these make use of them?

Sure, gave that a shot, LMK what you think.

> Also, it might be useful to have JavaDoc on these methods, simply to help future developers understand the intent.

Gave this a try too.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -156,6 +165,14 @@ public KafkaBasedLog(String topic,
+        // as it will not take records from currently-open transactions into account. We want to err on the side of caution in that
+        // case: when users request a read to the end of the log, we will read up to the point where the latest offsets visible to the
+        // consumer are at least as high as the (possibly-part-of-a-transaction) end offsets of the topic.
+        this.requireAdminForOffsets = IsolationLevel.READ_COMMITTED.name().toLowerCase(Locale.ROOT)
+                .equals(consumerConfigs.get(ConsumerConfig.ISOLATION_LEVEL_CONFIG));
```
Why not `equalsIgnoreCase(...)` here instead of lowercasing and then calling `equals(...)`?

C0urante replied:
This was my first thought, but the `isolation.level` property is case-sensitive.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -118,35 +123,39 @@ public KafkaBasedLog(String topic,
+     * @param initializer               the function that should be run when this log is {@link #start() started}; may be null
+     * @param topicContainsTransactions whether the topic being consumed contains (or is expected to contain) transactions;
+     *                                  if this is {@code false} and the topic does contain transactions, reads to the end of the log may block
+     *                                  indefinitely
```
There are projects outside of Apache Kafka that do use this class. While this class is not a public API and we technically don't have to avoid breaking compatibility, it's fairly straightforward to maintain API compatibility and so we should do that.

C0urante replied:
I'll rebase this PR on top of https://github.com/apache/kafka/pull/11046 and, after removing any changes to this class that become unnecessary as a result, will address this.

I do believe a KIP is called for here though. As you've noted, this class is not public API and it's frustrating that development efforts against Connect are hampered by this unofficial and somewhat arbitrary restriction.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -210,26 +227,21 @@ public void stop() {
-            log.error("Failed to stop KafkaBasedLog consumer", e);
-        }
+        Utils.closeQuietly(producer, "KafkaBasedLog producer");
+        Utils.closeQuietly(consumer, "KafkaBasedLog consumer");
```
This change does alter the level at which these problems are logged. Is that intentional, and if so why?

C0urante replied:
Fewer lines, more standardized, and logging these at `ERROR` level is incorrect IMO.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/ConnectUtils.java:
```diff
@@ -72,6 +74,45 @@ static String lookupKafkaClusterId(Admin adminClient) {
+            ));
+        } else {
+            return Optional.empty();
+        }
```
Nit: what do you think about tolerating an empty or null `justification` string, since this methods does not ensure that one is provided?

C0urante replied:
👍  &nbsp; SGTM

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java:
```diff
@@ -43,6 +44,13 @@
+ * over to the secondary backing store on a best-effort basis. Failures to write to the secondary
+ * store are logged but otherwise swallowed silently.
+ * </p>
+ * <p>
```
Did you consider introducing a new `OffsetBackingStore` implementation that writes to two other `OffsetBackingStore` implementations? That might simplify the logic in this class and better encapsulate the double write behavior.

C0urante replied:
This is a neat idea. It'd also simplify the offset read logic. I've taken a stab at this; LMK what you think.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java:
```diff
@@ -114,12 +124,18 @@ public synchronized boolean beginFlush() {
         if (data.isEmpty())
             return false;
 
-        assert !flushing();
```
Why remove this assertion?

C0urante replied:
It's incorrect; this is a known issue where a race condition involving task failure during offset commit can cause this assertion to fail. The assertion itself has provided no value in catching, reproducing, or testing against this issue and at this point is more misleading than useful.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetStorageWriter.java:
```diff
@@ -163,13 +181,24 @@ public synchronized boolean beginFlush() {
+                if (secondaryBackingStore != null && primaryError == null) {
+                    secondaryBackingStore.set(offsetsSerialized, (secondaryError, secondaryResult) -> {
+                        if (secondaryError != null) {
+                            log.warn("Failed to write offsets ({}) to secondary backing store", flushed, secondaryError);
+                        } else {
+                            log.debug("Successfully flushed offsets ({}) to secondary backing store", flushed);
```
Do these log messages include the Connector context for EOS-enabled source connectors?

C0urante replied:
Good catch; I believe there's also a case in the existing case where offset commit messages that are logged in a producer callback are also missing the MDC context. I've addressed both cases.

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceTask.java:
```diff
@@ -20,13 +20,45 @@
 public abstract class SourceTask implements Task {
 
+    /**
+     * <p>
```
We don't need this `<p>` tag.

C0urante replied:
Ack, removed.

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceTask.java:
```diff
@@ -20,13 +20,45 @@
+     * <p>
+     * The configuration key that determines how source tasks will define transaction boundaries
+     * when exactly-once support is enabled.
+     * </p>
```
The `</p>` has no meaning in JavaDoc, so we should remove these.

C0urante replied:
Ack, removed.

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/TransactionContext.java:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * Request a transaction commit after a source record is processed. The source record will be the
+     * last record in the committed transaction.
+     * @param record the record to commit the transaction after.
```
Is the SourceTask implementation calling this method allowed to pass a null value here? If so, what happens? If not, please add that to the JavaDoc.

C0urante replied:
Ack, added "may not be null".

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/TransactionContext.java:
```diff
@@ -0,0 +1,55 @@
+     * and will not appear in a committed transaction. However, offsets for that transaction will still
+     * be committed. If the data should be reprocessed, the task should not invoke this method and
+     * should instead throw an exception.
+     * @param record the record to abort the transaction after.
```
Is the SourceTask implementation calling this method allowed to pass a null value here? If so, what happens? If not, please add that to the JavaDoc.

C0urante replied:
Ack, added "may not be null".

rhauch reviewed (commented):
Thanks, @C0urante. I took a first pass over a portion of the PR, and tried to focus mostly on the public API portions. See comments below. I plan to take another pass on other parts soon.

C0urante commented:
Resolved merge conflicts. Will try to address review comments soon.

C0urante commented:
@rhauch Is that second pass coming any time soon? It'd be nice if we could get this merged in time for the upcoming 3.1 release.

I plan to address the existing comments next week.

C0urante commented:
Rebased onto https://github.com/apache/kafka/pull/11046 and cleaned up the `KafkaBasedLog` changes to remove the ugly hack for handling aborted transactions.

C0urante commented:
Removed from the upcoming 3.1 release plan due to lack of review.

@rhauch do you think we can get this merged in time for the 3.2/4.0 release? I'd rather not have to continue cleaning up merge conflicts for this PR and the KIP was accepted several months ago with only one round of review since then. If there just isn't time to look at this please let me know and I can close the PR.

rhauch commented:
Thanks, @C0urante. I have been trying to get through this large PR, but that takes time and I often get sidetracked by other things, such as AK 3.1 release issues.

Please keep this open. I do plan to complete my latest pass of this PR soon. Thanks for your continued patience.

C0urante commented:
@rhauch gentle reminder, it's been another month and we're still waiting on that next pass.

galeaspablo commented:
Hey folks, I've kept an eye on this PR, its associated KIP, for a while. Is there anything I can do to help?

Happy to take guidance to split the PR if that would make reviewing easier?

P.S. I appreciate I haven't committed to this repository, but all my Kafka work was internal while I was at AWS.

galeaspablo commented:
Hey @rhauch, @C0urante, I've kept an eye on this PR, its associated KIP, for a while. Is there anything I can do to help? Happy to take guidance to split the PR if that would make reviewing easier?

P.S. I appreciate I haven't committed to this repository, but all my Kafka work was internal while I was at AWS.

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceTaskContext.java:
```diff
@@ -38,4 +38,29 @@
+     * </pre>
+     *
+     * @return the transaction context, or null if the connector was not configured to specify transaction boundaries
+     * @since 3.0
```
Please change all of these `@since 3.0` to `@since 3.1`, plus any other `Kafka 3.0` references in JavaDoc.

C0urante replied:
I believe 3.2 is the correct version now, right? Will update to that.

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java:
```diff
@@ -28,4 +30,39 @@
+    /**
+     * Signals whether the connector supports exactly-once delivery guarantees with a proposed configuration.
+     * Developers can assume that worker-level exactly-once support is enabled when this method is invoked.
+     * For backwards compatibility, the default implementation will return {@code null}, but connector developers are
```
Nit: using a new paragraph makes this stand out more.
```suggestion
     * <p>For backwards compatibility, the default implementation will return {@code null}, but connector developers are
```

C0urante replied:
Ack, done. (And added a few more paragraph breaks in other places too).

rhauch commented on connect/api/src/main/java/org/apache/kafka/connect/source/TransactionContext.java:
```diff
@@ -0,0 +1,55 @@
+ */
+package org.apache.kafka.connect.source;
+
+/**
+ * Provided to source tasks to allow them to define their own producer transaction boundaries when
+ * exactly-once support is enabled.
+ */
+public interface TransactionContext {
```
IIUC, the `WorkerTransactionContext` is the only implementation of this. That means that if a connector is configured with `transaction.boundary=poll` or `transaction.boundary=interval`, a poorly-implemented connector could still call these methods and they'd unnecessarily accumulate records.

WDYT about in such cases the `SourceTaskContext#transactionContext()` method returning a no-op implementation of this interface, so no harm is done if a connector implementation still calls these methods when `transaction.boundary` is _not_ set to `connector`?

Maybe we could consider a warning log message if these methods are called by a connector inappropriately. But we have to be careful. While such log messages might be useful for the **developer** of a connector plugin, I would argue that _prolific_ warnings are actually harmful for a **user** trying to _use_ a connector plugin they didn't develop with a connector configuration that includes `transaction.boundary=poll` or `transaction.boundary=interval`. So maybe it's worthwhile for the "no-op" implementation to only log each warning once per method per instance.

C0urante replied:
The behavior specified in the KIP is for `SourceTaskContext::transactionContext` to return "the transaction context, or null if the connector was not configured to specify transaction boundaries" (see the Javadocs for that method in the code snippets in [the relevant KIP section](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors#KIP618:ExactlyOnceSupportforSourceConnectors-ConnectorAPIexpansions)).

This is implemented in the [`ExactlyOnceWorkerSourceTask`](https://github.com/C0urante/kafka/blob/c06207d64a76286316b862e16143f909181501c3/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java#L113-L117) class and should prevent connectors from invoking `TransactionContext` methods and accumulating records when the user has not configured the connector to define its own transaction boundaries.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedConfig.java:
```diff
@@ -192,6 +199,44 @@
+    }
+
+    public static final String EXACTLY_ONCE_SOURCE_SUPPORT_CONFIG = "exactly.once.source.support";
+    public static final String EXACTLY_ONCE_SOURCE_SUPPORT_DOC = "Whether to enable exactly-once support for source connectors in the cluster "
+            + "by writing source records and their offsets in a Kafka transaction, and by proactively fencing out old task generations before bringing up new ones. "
```
Nit, to improve readability and precision, especially around how many Kafka transactions would be used:
> Whether to enable exactly-once support for source connectors in the cluster by using transactions to write source records and their source offsets, and by proactively fencing out old task generations before bringing up new ones.

C0urante replied:
Ack, done.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedConfigTest.java:
```diff
@@ -294,4 +298,41 @@ public void shouldRemoveCompactionFromStatusTopicSettings() {
+                "connect-cluster-\u2603",
+                new DistributedConfig(workerProps).transactionalProducerId()
+        );
+    }
```
Let's add some negative tests that verify that invalid values for the new worker configuration properties are properly handled/identified by the validators.

We should also add positive and negative tests for the new connector-level config properties in  `SourceConnectorConfigTest`

C0urante replied:
Ack, done. The tests are located elsewhere since the logic is implemented in a different class, but there are new cases for this in the `KafkaConfigBackingStoreTest`, `KafkaOffsetBackingStoreTest`, and `WorkerTest` test suites.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -125,28 +130,28 @@ public KafkaBasedLog(String topic,
      *                           setting, such as the auto offset reset policy, will be overridden to ensure correct
      *                           behavior of this class.
-     * @param topicAdminSupplier supplier function for an admin client, the lifecycle of which is expected to be controlled
+     * @param topicAdminSupplier        supplier function for an admin client, the lifecycle of which is expected to be controlled
      *                           by the calling component; may not be null
-     * @param consumedCallback   callback to invoke for each {@link ConsumerRecord} consumed when tailing the log
-     * @param time               Time interface
-     * @param initializer        the function that should be run when this log is {@link #start() started}; may be null
+     * @param consumedCallback          callback to invoke for each {@link ConsumerRecord} consumed when tailing the log
+     * @param time                      Time interface
+     * @param initializer               the function that should be run when this log is {@link #start() started}; may be null
      */
     public KafkaBasedLog(String topic,
-            Map<String, Object> producerConfigs,
-            Map<String, Object> consumerConfigs,
-            Supplier<TopicAdmin> topicAdminSupplier,
-            Callback<ConsumerRecord<K, V>> consumedCallback,
-            Time time,
-            java.util.function.Consumer<TopicAdmin> initializer) {
+                         Map<String, Object> producerConfigs,
+                         Map<String, Object> consumerConfigs,
+                         Supplier<TopicAdmin> topicAdminSupplier,
+                         Callback<ConsumerRecord<K, V>> consumedCallback,
+                         Time time,
+                         java.util.function.Consumer<TopicAdmin> initializer) {
```
Why make these unnecessary changes?

C0urante replied:
Apologies, will revert. Believe this was left in from an older approach that involved more invasive changes to this class.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -365,6 +413,10 @@ private void readToLogEnd() {
                 // This may happen with really old brokers that don't support the auto topic creation
                 // field in metadata requests
                 log.debug("Reading to end of log offsets with consumer since admin client is unsupported: {}", e.getMessage());
+                if (requireAdminForOffsets) {
+                    // Should be handled by the caller during log startup
+                    throw e;
+                }
```
It's true that the `DistributedHerder.run()` is ultimately catching and handling this exception. But I feel like many users might not understand the significance of such an error nor how to correct their configuration. Rather than just re-throw that exception, we should probably wrap that exception with one that has a more instructive message, such as something like:
> Enabling exactly once for source connectors requires a Kafka broker version that allows admin clients to read consumer offsets. Disable the worker's exactly once support for source connectors, or use a newer Kafka broker version.

Plus, should this if block be before the `log.debug(...)` on the previous line? Seems like that log message might just confuse the situation since the worker will not read "to the end of log offsets with consumer".

C0urante replied:
Ack on both points; done 👍

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -264,12 +279,62 @@ public void startConnector(
+                final CloseableOffsetStorageReader offsetReader;
+                if (ConnectUtils.isSinkConnector(connector)) {
+                    connConfig = new SinkConnectorConfig(plugins, connProps);
+                    offsetReader = null;
```
IIUC, this changes the behavior of the `WorkerConnector` created below. Prior to this PR, the `WorkerConnector` was always created with the `Worker.offsetBackingStore`, even for sink connectors. 

However, with this PR, the `WorkerConnector` will be instantiated with a null `offsetReader` parameter, which will cause a NPE in `WorkerConnector#doShutdown()` and `WorkerConnector#cancel()`  since `WorkerConnector` does not check for a null parameter there.

C0urante replied:
Ah, good catch! I think this is one more place where `Utils::closeQuietly` comes in handy since it does the null check for us, won't interrupt shutdown if an exception is thrown, and further standardizes the cleanup logic for the code base.

Alternatively, we could retain the current behavior, but I don't see much use in constructing an offset reader for sink connectors that won't be used now and is unlikely to be used in the future.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java:
```diff
@@ -0,0 +1,195 @@
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+public class ConnectorOffsetBackingStore implements OffsetBackingStore {
```
Please add JavaDoc that explains the purpose of this class, and in particular, what the two modes are that correspond to the two static factory methods.

C0urante replied:
Ack, done. Added docs to the (now-three) static factory methods and to each of the overridden `OffsetBackingStore` methods.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java:
```diff
@@ -0,0 +1,195 @@
+    private final Optional<OffsetBackingStore> connectorStore;
+    private final Optional<TopicAdmin> connectorStoreAdmin;
+
+    public static ConnectorOffsetBackingStore withConnectorOffsetStore(
```
Nit: move these two static factory methods above the non-static member variables, so all static and non-static members are together.

C0urante replied:
Ack, done.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/ConnectUtils.java:
```diff
@@ -72,6 +74,45 @@ static String lookupKafkaClusterId(Admin adminClient) {
+        if (value != null && !matchesExpectedValue.test(value)) {
+            return Optional.of(String.format(
+                    "The value '%s' for the '%s' property will be ignored as it cannot be overridden%s. "
+                            + "The value '%s' will be used instead.",
```
Is this really true that all of these incorrectly-overridden property values are ignored?

C0urante replied:
I suppose not! There are two classes that currently leverage this utility method: `Worker` and `DistributedConfig`.

In the former, each invocation of `warnOnOverriddenProperty` is followed closely (if not immediately) by a line that overrides any user-supplied values that don't match the expected value.

However, in `DistributedConfig`, warnings are emitted for values that are ignored and overridden when constructing Kafka clients for some but not all `KafkaBasedLog` instances: specifically, the ones used for the config and offset backing stores. It's possible that a user might specify `enable.idempotence=false` in their worker config with the intention of disabling idempotent writes for the producers used for the status backing store (since the default for this property was updated from `false` to `true` in [KIP-679](https://cwiki.apache.org/confluence/display/KAFKA/KIP-679%3A+Producer+will+enable+the+strongest+delivery+guarantee+by+default)). With the current behavior for this PR, that change would have an effect (and the values logged for the producer for the status backing store would confirm this), but the warning message would still be logged, which might lead to some confusion.

At the bare minimum we can and should update this message to be more specific about where exactly these values will be ignored, but in addition to that, I've taken a stab at restructuring the reusable utility function to actually perform the override in addition to just warning about it, like you suggested in a comment below. This makes it harder to log warnings about overrides without actually ensuring that the overrides have taken place.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -264,12 +279,62 @@ public void startConnector(
+                } else {
+                    SourceConnectorConfig sourceConfig = new SourceConnectorConfig(plugins, connProps, config.topicCreationEnable());
+                    connConfig = sourceConfig;
+
```
Suggestion:
```suggestion

                    // Use the desired topic for offsets
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -264,12 +279,62 @@ public void startConnector(
+                    } else if (config.exactlyOnceSourceEnabled()) {
+                        connectorOffsetsTopic = config.offsetsTopic();
+                    }
+
```
Suggestion:
```suggestion

                    // Set up the offset backing store for this connector instance
```

C0urante replied:
Ack, done.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -476,22 +541,95 @@ public boolean isRunning(String connName) {
+    }
+
+    /**
+     * Start a source task managed by this worker.
```
It's probably worth while to mention that this method starts the task for a source connector with older behavior (without exactly once support).

C0urante replied:
Ack, done.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -692,21 +823,22 @@ private WorkerTask buildWorkerTask(ClusterConfigState configState,
     }
 
-    static Map<String, Object> adminConfigs(ConnectorTaskId id,
+    static Map<String, Object> adminConfigs(String connName,
```
Do we need to change these signatures from `ConnectorTaskId` to `String`? The `ConnectorTaskId` gives us the ability to define tasks-specific client configuration properties if necessary/desired. I'm afraid that switching to `String` will make it harder and more invasive to add that back in. Plus, if there's not a good reason to remove these, let's leave that for smaller PRs.

C0urante replied:
We have to be able to construct admin clients for `Connector` instances now, whereas before we only used them for `Task` instances.

I do sympathize with the desire for task-specific tweaks, though. Right now the pattern is for the `adminConfigs` method to define a new parameter for anything that should be derived from the task ID/connector name and place the onus on the caller to derive that value (for example, this is done with the `defaultClientId` parameter). I think this works for now but one pattern we could consider is to add separate wrapper `adminConfigs` methods for clients constructed on behalf of connectors and on behalf of tasks, similar to the new wrapper methods I've added for (regular and exactly-once) offset consumers and exactly-once task producers.

Thoughts?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -678,10 +810,9 @@ private WorkerTask buildWorkerTask(ClusterConfigState configState,
-                  Utils.join(config.getList(WorkerConfig.BOOTSTRAP_SERVERS_CONFIG), ","));
+        consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, SinkUtils.consumerGroupId(connName));
+        consumerProps.put(ConsumerConfig.CLIENT_ID_CONFIG, defaultClientId);
+        consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, config.bootstrapServers());
```
Thanks! This should have been done quite some time ago.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -980,6 +1119,329 @@ WorkerMetricsGroup workerMetricsGroup() {
+            );
+            final TopicAdmin admin;
+            Map<String, TopicCreationGroup> topicCreationGroups = null;
+            if (sourceConfig.offsetsTopic() != null || (config.topicCreationEnable() && sourceConfig.usesTopicCreation())) {
```
I think adding some comments here will help future developers. There are enough subtleties here (e.g., the `||` in this line) that makes this more challenging than it should be to understand the behavior. 

Or, it might be worth adding some final booleans here to make these conditions a bit more clear and readable, especially since this method is a series of relatively independent parts/sections. For example:
```
final boolean customOffsetTopic = sourceConfig.offsetTopic();
final boolean createTopicsEnabled = config.topicCreationEnable() && sourceConfig.usesTopicCreation();
if (customOffsetTopic || createTopicsEnabled) {
  // Create an admin client
  ...
  if (createTopicsEnabled) {
    topicCreationGroups = TopicCreationGroup.configuredGroups(sourceConfig);
  }
  if (customOffsetTopic) {
    // Build custom offset store
    ...
  }
} else {
  // No need for admin
  admin = null;
}
```

C0urante replied:
I've extracted a lot of this logic into separate methods for testability and readability, but tried to stick with the general spirit of this comment while doing so and left a few comments and added some `final boolean` local variables in the newly-isolated methods.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -980,6 +1119,329 @@ WorkerMetricsGroup workerMetricsGroup() {
+            TransformationChain<SourceRecord> transformationChain = new TransformationChain<>(sourceConfig.<SourceRecord>transformations(), retryWithToleranceOperator);
+            log.info("Initializing: {}", transformationChain);
+
+            Map<String, Object> adminOverrides = adminConfigs(id.connector(), "connector-adminclient-" + id, config,
```
In my comment about `SourceTaskBuilder#doBuild`, I suggested adding a few comments to help the different parts/sections of this method stand out more clearly. The same would be true here, such as:
```suggestion
            // Create an admin client
            Map<String, Object> adminOverrides = adminConfigs(id.connector(), "connector-adminclient-" + id, config,
```

C0urante replied:
This section has been pretty heavily refactored; I hope the current state of things honors the intent of this comment but if not LMK.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -980,6 +1119,329 @@ WorkerMetricsGroup workerMetricsGroup() {
+            String transactionalId = transactionalId(id);
+            Map<String, Object> producerProps = producerConfigs(id.connector(), "connector-producer-" + id, config, sourceConfig, connectorClass,
+                    connectorClientConfigOverridePolicy, kafkaClusterId);
+            ConnectUtils.warnOnOverriddenProperty(
+                    producerProps, ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "true",
+                    "for connectors when exactly-once source support is enabled",
+                    false
+            );
+            ConnectUtils.warnOnOverriddenProperty(
+                    producerProps, ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId,
+                    "for connectors when exactly-once source support is enabled",
+                    true
+            );
+            producerProps.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
+            producerProps.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId);
```
Ah, so here's one example of how these properties cannot be overridden by the connector config. But that's not quite so obvious for some other calls to `ConnectUtils.warnOnOverriddenProperty`.

Is it not feasible to reset the expected value on the props within the `ConnectUtils#warnOnOverriddenProperty` method?

C0urante replied:
Yep, done.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -980,6 +1119,329 @@ WorkerMetricsGroup workerMetricsGroup() {
+            String offsetsTopic = Optional.ofNullable(sourceConfig.offsetsTopic()).orElse(config.offsetsTopic());
+
+            ConnectorOffsetBackingStore offsetStore;
+            // No need to do secondary writes to the global offsets topic if we're certain that the task's local offset store
+            // is going to be targeting it anyway
+            // Note that this may lead to a false positive if the user provides an overridden bootstrap servers value for their
+            // producer that resolves to the same Kafka cluster; we might consider looking up the Kafka cluster ID in the future
+            // to prevent these false positives but at the moment this is probably adequate, especially since we probably don't
+            // want to put a ping to a remote Kafka cluster inside the herder's tick thread (which is where this logic takes place
+            // right now) in case that takes a while.
```
This is more of a nit-type comment.

It seems like this comment tries to describe why we don't need to create a connector-specific offset store **AND** then talk about the check might not always be accurate. But I found the comment a bit hard to follow.

Also, the `ConnectorOffsetBackingStore offsetStore;` line should move under the multi-line comment.

In such cases, it might be worth changing the code to help make things more clear. For example:
```
// We can simply reuse the worker's offset store when the connector-specific offset topic
// is the same as the worker's. We can check the offset topic name and the Kafka cluster's
// bootstrap servers, although this isn't exact and can lead to some false positives if the user
// provides an overridden bootstrap servers value for their producer that is different than
// the worker's but still resolves to the same Kafka cluster used by the worker.
// At the moment this is probably adequate, especially since we probably don't want to put
// a network ping to a remote Kafka cluster inside the herder's tick thread (which is where this
// logic takes place right now) in case that takes a while.
ConnectorOffsetBackingStore offsetStore;
final boolean sameOffsetTopicAsWorker = offsetsTopic.equals(config.offsetsTopic())
                    && producerProps.get(BOOTSTRAP_SERVERS_CONFIG).equals(config.bootstrapServers();
if (sameOffsetTopicAsWorker) {
    offsetStore = ConnectorOffsetBackingStore.withoutConnectorOffsetStore(
    ...
} else {
    offsetStore = ConnectorOffsetBackingStore.withConnectorOffsetStore(
    ...
}
```

C0urante replied:
Ack, done. This logic has been moved into an isolated method and I've moved the relevant portions of the comment with it.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:
```diff
@@ -342,6 +343,10 @@ private void logPluginPathConfigProviderWarning(Map<String, String> rawOriginals
         }
     }
 
+    public String bootstrapServers() {
+        return String.join(",", getList(BOOTSTRAP_SERVERS_CONFIG));
+    }
```
Very nice. For this and the other new getter methods, can you add some JavaDoc just so that it's easier to follow in an IDE where these methods are _used_? I wanted that several times as I was reviewing this code.

C0urante replied:
Sorry, I'm a little unclear on what the benefits of Javadocs here are. I've added tags for each of the relevant methods but would you mind elaborating? In IntelliJ, I can see where this method is invoked by cmd+clicking the `bootstrapServers()` declaration, and wherever it's invoked, I can go to the declaration by cmd+clicking the invocation.

I'm aware that not everyone uses my IDE and certainly don't expect them to start doing so; just wondering if there's a different IDE out there with behavior that makes Javadocs like these significantly more powerful.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -74,15 +75,15 @@
                            ConnectMetrics metrics,
                            ConnectorStatus.Listener statusListener,
-                           OffsetStorageReader offsetStorageReader,
+                           CloseableOffsetStorageReader offsetStorageReader,
```
As mentioned above, IIUC this PR will sometimes pass a null `offsetStorageReader` (IIRC for sink connectors), but this class currently expects that to be null. Might be worth adding a `Objects.requireNonNull(...)` call here to help catch that situation.

C0urante replied:
I opted for an alternative approach where the offset reader is allowed to be `null`, but I can also see the benefit of a conditional call to `Objects::requireNonNull`. I've added one in `WorkerConnector::initialize`; let me know what you think.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java:
```diff
@@ -16,95 +16,46 @@
  * WorkerTask that uses a SourceTask to ingest data into Kafka.
  */
-class WorkerSourceTask extends WorkerTask {
+class WorkerSourceTask extends AbstractWorkerSourceTask {
```
Note to future me: I didn't get this far in the PR.

C0urante replied:
May be best to hold off on this section for a bit; there's a fair amount of rebasing that needs to be done now that https://github.com/apache/kafka/pull/11323 has been merged.

rhauch reviewed (commented):
Thanks for the updates, @C0urante. And I apologize for the delay, which was due to other commitments.

I've reviewed most of the public API portions, and a fair portion of the `Worker` and other similar classes. I've not yet made it through the refactoring of the `WorkerSourceTask`. I've slotted time next week to keep going.

(Also, I responded on a few threads from my previous review. I resolved most of that review's threads where I was happy with your fixes. Hopefully that makes it easier to see the outstanding threads.)

C0urante commented:
Thanks Randall, I've addressed all of your comments either with code changes or an inline response in the appropriate thread.

I've noted this elsewhere but it bears repeating: the `WorkerSourceTask` and `AbstractWorkerSourceTask` classes may be best to save for later given that there are some conflicts that will need to be resolved between this PR and the current `trunk` branch.

The `ExactlyOnceWorkerSourceTask` class should be fine for review, as well as the zombie fencing logic added to the `DistributedHerder`, `KafkaConfigBackingStore` and `Worker` classes.

The new admin client API for fencing producer transactions should also be safe to review (although there are some conflicts in the `KafkaAdminClientTest` that need to be resolved).

> (Also, I responded on a few threads from my previous review. I resolved most of that review's threads where I was happy with your fixes. Hopefully that makes it easier to see the outstanding threads.)

This made it very easy to follow along, thanks 👍

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java:
```diff
@@ -0,0 +1,507 @@
+
+        long started = time.milliseconds();
+
+        // Necessary as a workaround for https://issues.apache.org/jira/browse/KAFKA-9279,
+        // where some calls to Producer::send fail asynchronously but do not cause a subsequent call
+        // to Producer::commitTransaction to throw an exception
+        // We can remove this once a fix for KAFKA-9279 is merged
+        maybeThrowProducerSendException();
```
I've filed https://github.com/apache/kafka/pull/11508 as a fix for [KAFKA-9279](https://issues.apache.org/jira/browse/KAFKA-9279). The workaround here isn't terribly awkward so it probably isn't necessary for us to block on that fix PR, but it'd be nice if someone could find the time to review it.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTask.java:
```diff
@@ -0,0 +1,643 @@
+                        if (e != null) {
+                            log.debug("{} failed to send record to {}: ", AbstractWorkerSourceTask.this, topic, e);
+                            log.trace("{} Failed record: {}", AbstractWorkerSourceTask.this, preTransformRecord);
+                            if (isFatalProducerSendException(e)) {
+                                producerSendException.compareAndSet(null, e);
+                            }
```
This will have to be tweaked slightly in order to invoke `SourceTask::commitRecord` and correctly track the number of in-flight records for metrics. Leaving as-is for now since this part of the PR will have to be updated anyways once rebased onto `trunk`.

mimaison commented:
@C0urante Thanks for the PR! It's really massive! Can we try to cut it in smaller chunks? That would make it a lot easier to review and allow merging bit by bit.

C0urante commented:
I've completed a rebase onto the latest trunk, which brought in notable changes from https://github.com/apache/kafka/pull/11323, https://github.com/apache/kafka/pull/11382, https://github.com/apache/kafka/pull/11508 (thanks Mikael!), and https://github.com/apache/kafka/pull/11046.

I'll start to break things up into smaller pieces tomorrow.

C0urante commented:
@mimaison @rhauch I've broken things down into smaller PRs.

These can all be reviewed immediately:
- https://github.com/apache/kafka/pull/11777 (cc @hachikuji in case you want to take a look)
- https://github.com/apache/kafka/pull/11772
- https://github.com/apache/kafka/pull/11773
- https://github.com/apache/kafka/pull/11774

These have requirements on other PRs (either ones outlined above, or ones also in this list):
- https://github.com/apache/kafka/pull/11775
- https://github.com/apache/kafka/pull/11776
- https://github.com/apache/kafka/pull/11778
- https://github.com/apache/kafka/pull/11779
- https://github.com/apache/kafka/pull/11780
- https://github.com/apache/kafka/pull/11781

And these should be reviewed last as they bring all of the above changes together and add integration and system tests for them, respectively:
- https://github.com/apache/kafka/pull/11782
- https://github.com/apache/kafka/pull/11783
